test/io/test_filesystem.f90

```diff
@@ -0,0 +1,258 @@
+        call check(error, stat, "Listing the contents of an empty directory shouldn't fail.")
+        call check(error, size(files) == 1, "The directory should contain one file.")
+        call check(error, char(files(1)) == filename, "The file should be '"//filename//"'.")
+
+        call run('rm -rf '//temp_list_dir, iostat=stat)
+    end
+
+    subroutine fs_list_dir_two_files(error)
+        type(error_type), allocatable, intent(out) :: error
+
+        integer :: stat
+
+        type(string_type), allocatable :: files(:)
+        character(*), parameter :: filename1 = 'abc.txt'
+        character(*), parameter :: filename2 = 'xyz'
+
+        call run('rm -rf '//temp_list_dir, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Removing directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('mkdir '//temp_list_dir, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Creating directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('touch '//temp_list_dir//'/'//filename1, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Creating file 1 in directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('touch '//temp_list_dir//'/'//filename2, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Creating file 2 in directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call list_dir(temp_list_dir, files, stat)
+        call check(error, stat, "Listing the contents of an empty directory shouldn't fail.")
+        call check(error, size(files) == 2, "The directory should contain two files.")
+        call check(error, char(files(1)) == filename1, "The file should be '"//filename1//"'.")
+        call check(error, char(files(2)) == filename2, "The file should be '"//filename2//"'.")
+
+        call run('rm -rf '//temp_list_dir, iostat=stat)
+    end
+
+    subroutine fs_list_dir_one_file_one_dir(error)
+        type(error_type), allocatable, intent(out) :: error
+
+        integer :: stat
+
+        type(string_type), allocatable :: contents(:)
+        character(*), parameter :: filename1 = 'abc.txt'
+        character(*), parameter :: dir = 'xyz'
+
+        call run('rm -rf '//temp_list_dir, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Removing directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('mkdir '//temp_list_dir, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Creating directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('touch '//temp_list_dir//'/'//filename1, iostat=stat)
+        if (stat /= 0) then
+            call test_failed(error, "Creating file 1 in directory '"//temp_list_dir//"' failed."); return
+        end if
+
+        call run('mkdir '//temp_list_dir//'/'//dir, iostat=stat)
```

Comment:
[There is an error in CI under Windows OS](https://github.com/fortran-lang/stdlib/actions/runs/10888590749/job/30213419329?pr=874#step:7:1004). This is because the shell command `mkdir` called may follow the `mkdir` command in `CMD.exe`, like `mkdir \a\b\c\d`. The path connector uses `\` instead of `/`. It may be possible to implement a routine to determine the current operating system in Stdlib later.

A temporary solution can be to make two judgments. If the `/` scheme fails, then try the `\` scheme.

Reply:
A simple solution that I've been using and think would be useful to have in stdlib would be to add in the constants module something like:
```Fortran
#ifdef _WIN32
  character(len=1), parameter, public :: os_sep='\'
#else
  character(len=1), parameter, public :: os_sep='/'
#endif
```
Which then would enable the equivalent of python
```python
import os
print(os.sep)
```

Reply:
The C preprocessor **in Fortran** does not predefine the `_WIN32` macro definition. This may require CMake to determine the operating system and define `_WIN32`. There are indeed two solutions:
1. During compilation, solve this problem through the C preprocessor and CMake macro definitions.
2. During runtime, solve this problem by implementing a routine to determine the operating system.

Reply:
The C preprocessor approach is already in use in stdlib_system, so it could be one solution. The runtime approach is also very good.

Reply:
I chose the runtime approach because, as mentioned by @zoziha, Windows-GFortran does not set the `_WIN32` macro. We had that problem in `fpm` already and I don't think the solution in `stdlib_system` works on Windows using `gfortran`.